src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    /// satisfying the original request. This condition implies that
+    /// such an allocation request will never succeed on the given
+    /// allocator, regardless of environment, memory pressure, or
+    /// other contextual condtions.
+    ///
+    /// For example, an allocator that does not support zero-sized
+    /// blocks can return this error variant.
+    Unsupported { details: &'static str },
+}
+
+impl AllocErr {
+    pub fn invalid_input(details: &'static str) -> Self {
+        AllocErr::Unsupported { details: details }
+    }
+    pub fn is_memory_exhausted(&self) -> bool {
+        if let AllocErr::Exhausted { .. } = *self { true } else { false }
+    }
+    pub fn is_request_unsupported(&self) -> bool {
+        if let AllocErr::Unsupported { .. } = *self { true } else { false }
+    }
+}
+
+/// The `CannotReallocInPlace` error is used when `fn realloc_in_place`
+/// was unable to reuse the given memory block for a requested layout.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct CannotReallocInPlace;
+
+/// An implementation of `Allocator` can allocate, reallocate, and
+/// deallocate arbitrary blocks of data described via `Layout`.
+///
+/// Some of the methods require that a layout *fit* a memory block.
+/// What it means for a layout to "fit" a memory block means is that
+/// the following two conditions must hold:
+///
+/// 1. The block's starting address must be aligned to `layout.align()`.
+///
+/// 2. The block's size must fall in the range `[use_min, use_max]`, where:
+///
+///    * `use_min` is `self.usable_size(layout).0`, and
+///
+///    * `use_max` is the capacity that was (or would have been)
+///      returned when (if) the block was allocated via a call to
+///      `alloc_excess` or `realloc_excess`.
+///
+/// Note that:
+///
+///  * the size of the layout most recently used to allocate the block
+///    is guaranteed to be in the range `[use_min, use_max]`, and
+///
+///  * a lower-bound on `use_max` can be safely approximated by a call to
+///    `usable_size`.
+///
+pub unsafe trait Allocator {
+
+    /// Returns a pointer suitable for holding data described by
+    /// `layout`, meeting its size and alignment guarantees.
+    ///
+    /// The returned block of storage may or may not have its contents
+    /// initialized. (Extension subtraits might restrict this
+    /// behavior, e.g. to ensure initialization.)
+    ///
+    /// Returning `Err` indicates that either memory is exhausted or `layout` does
+    /// not meet allocator's size or alignment constraints.
+    ///
+    /// Implementations are encouraged to return `Err` on memory
+    /// exhaustion rather than panicking or aborting, but this is
+    /// not a strict requirement. (Specifically: it is *legal* to use
+    /// this trait to wrap an underlying native allocation library
+    /// that aborts on memory exhaustion.)
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<Address, AllocErr>;
```

Comment:
Hmm. I personally don't think it is a burden to make `fn alloc` an unsafe method, since most clients will either need to immediately use unsafe code to initialize the allocated state, or will be passing along an uninitialized block of memory which means that they themselves will probably be unsafe fn's as well...

But I agree that its important to spell out all of the criteria for what safe input arguments are, e.g. requiring layouts have non-zero size...